source/grid/grid_tools.cc

```diff
@@ -4451,6 +4452,299 @@ namespace GridTools
+            };
+
+            auto cell_creation = [&](
+                                   const unsigned int vv0,
+                                   const unsigned int vv1,
+                                   const unsigned int f0,
+                                   const unsigned int f1,
+
+                                   const unsigned int n0,
+                                   const unsigned int v0n0,
+                                   const unsigned int v1n0,
+
+                                   const unsigned int n1,
+                                   const unsigned int v0n1,
+                                   const unsigned int v1n1)
+            {
+              CellData<dim> c1, c2;
+              CellData<1> l1, l2;
+
+              c1.vertices[v0] = cell->vertex_index(vv0);
+              c1.vertices[v1] = cell->vertex_index(vv1);
+              c1.vertices[v2] = cell->neighbor(n0)->vertex_index(v0n0);
+              c1.vertices[v3] = cell->neighbor(n0)->vertex_index(v1n0);
+
+              c1.manifold_id = cell->manifold_id();
+              c1.material_id = cell->material_id();
+
+              c2.vertices[v0] = cell->vertex_index(vv0);
+              c2.vertices[v1] = cell->neighbor(n1)->vertex_index(v0n1);
+              c2.vertices[v2] = cell->vertex_index(vv1);
+              c2.vertices[v3] = cell->neighbor(n1)->vertex_index(v1n1);
+
+              c2.manifold_id = cell->manifold_id();
+              c2.material_id = cell->material_id();
+
+              l1.vertices[0] = cell->vertex_index(vv0);
+              l1.vertices[1] = cell->neighbor(n0)->vertex_index(v0n0);
+
+              l1.boundary_id = cell->line(f0)->boundary_id();
+              l1.manifold_id = cell->line(f0)->manifold_id();
+              subcelldata_to_add.boundary_lines.push_back(l1);
+
+              l2.vertices[0] = cell->vertex_index(vv0);
+              l2.vertices[1] = cell->neighbor(n1)->vertex_index(v0n1);
+
+              l2.boundary_id = cell->line(f1)->boundary_id();
+              l2.manifold_id = cell->line(f1)->manifold_id();
+              subcelldata_to_add.boundary_lines.push_back(l2);
+
+              cells_to_add.push_back(c1);
+              cells_to_add.push_back(c2);
+            };
+
+            if (dim == 2)
+              {
+                switch (vertex_at_corner)
+                  {
+                  case 0:
+                    flags_removal(0,2,3,1);
+                    cell_creation(0,3, 0,2, 3,2,3,  1,1,3);
+                    break;
+                  case 1:
+                    flags_removal(1,2,3,0);
+                    cell_creation(1,2, 2,1, 0,0,2, 3,3,2);
+                    break;
+                  case 2:
+                    flags_removal(3,0,1,2);
+                    cell_creation(2,1, 3,0, 1,3,1, 2,0,1);
+                    break;
+                  case 3:
+                    flags_removal(3,1,0,2);
+                    cell_creation(3,0, 1,3, 2,1,0, 0,2,0);
+                    break;
```

Comment:
I also want to see if I can find a simpler approach that can replace these lambdas and the hardcoding...